A30-xds-v3.md

```diff
@@ -0,0 +1,255 @@
+any time, but none of the existing gRPC clients will actually use it
+unless you take additional steps.
+
+- For gRPC clients that support versions N and N+1, all you need to do
+  is add the server feature for version N+1 to the bootstrap file.  That
+  will cause these clients to start using version N+1.
+- For older gRPC clients that support versions N-1 and N, they will
+  ignore the server feature for version N+1 in the bootstrap file, so
+  they will continue to use version N.  You will need to upgrade to a
+  more recent release of gRPC to get them to use version N+1.
+
+Note that the only case in which users will need to upgrade their gRPC
+binaries in lock step with their xDS server is if they are migrating
+their xDS server from version N-1 to version N+1 in a single step,
+without adding support for version N in between.  We do not recommend
+this kind of transition.
+
+### Details of the v2 to v3 Transition
+
+The xDS major version number appears in two main places in the API:
+
+- The RPC method name used to make the xDS request.  For v2, this is
+  `/envoy.service.discovery.v2.AggregatedDiscoveryService/StreamAggregatedResources`,
+  whereas for v3 it is
+  `/envoy.service.discovery.v3.AggregatedDiscoveryService/StreamAggregatedResources`.
+  We call this the *transport protocol version*.
+- The type of resources that we request via the transport protocol.  For
+  example, the type for a Listener resource in v2 is `envoy.api.v2.Listener`,
+  whereas in v3 the type is `envoy.config.listener.v3.Listener`.  We
+  call this the *resource version*.
+
+When gRPC uses v3 to talk to a server, the primary thing that it is
+choosing is the transport protocol version, not the resource version.
+
+Luckily, we can be more flexible with the resource version, because
+there's essentially only one case where we use a field in the v2 protos
+that was removed in the v3 protos.  (That case is the node `build_version`
+field, which we can handle as a special case.)  Therefore, we can
+simply change our code to parse all incoming messages as v3, regardless
+of whether they are actually v2 or v3 messages.  We will also need to
+accept both v2 and v3 type names interchangeably in all `type_url` fields.
+
+This means that the logic for determining the transport protocol version
+is completely independent of the logic for handling the resource version.
+gRPC will always request resources of the same version as the transport
+protocol it is using.  However, it will accept any supported version of
+resources in responses from the server, regardless of what version of
+the transport protocol we are using.
+
+In other words, if we're using the v2 transport protocol, we will
+request v2 resources, but it's fine for the server to send us either v2
+or v3 resources.  Similarly, if we're using the v3 transport protocol,
+we will request v3 resources, but it's fine for the server to send us
+either v2 or v3 resources.
```

Comment:
Except that the server doesn't know that the client supports v2. When we add support for v3.1, we would drop support for these v2 resources and newer clients would break.

Reply:
Again, because of make-before-break, I think this is okay.  The client will still deserialize the resources as whatever version it wants, regardless of which version the server used when serializing.

Reply:
There is a mapping from v2 resource names to v3 resource names. You don't just look at the message name, "oh, this ends with `Listener`, let's assume it is an `envoy.config.listener.v3.Listener`". The version is part of the `Any` and `type_url`. When the client drops v2 support, are you saying it will keep the v2→v3 message name mapping and continue supporting v2 protos? That is not at all obvious to me.

Reply:
As above. The idea is that mixed version support only exists during the 1 year transition period and is only there for supporting rollouts where they may be mixed client/server features as an ephemeral process.

Reply:
I've added a paragraph clarifying that this is just intended for use as a rollout strategy, not something that can be assumed in the general case.